### ISSUE TITLE
Push a message to the driver upon worker death

### ISSUE THREAD

**Issue by stephanie-wang** [ray-project/ray#318]
When a worker dies, we provide availability by marking the task as lost and later reexecuting the task if the result is needed. However, if the worker death is a chronic problem related to the task itself, the workload will hang because every worker that tries to reexecute the task will die. For this reason, we should notify the users after (repeated) worker deaths, probably by pushing a message to the driver through Redis.

**Comment by robertnishihara** [comment]
A couple related things.

One is the PR #249 which enabled pushing error messages to specific drivers from C. It was closed, but something like that would be useful.

Second, we try to catch most exceptions that could happen on a worker process with by using `try` and `except` within the worker. If we missed something, then there is also a `try` `except` around the entire worker main loop, which will catch errors that are otherwise unhandled and will push an error to the relevant driver. https://github.com/ray-project/ray/blob/master/python/ray/workers/default_worker.py#L51

One thing to keep in mind is that when a worker dies, the task it was supposed to be executing presumably did not finish, so the task return values were not created. This means that other processes (a driver or other workers) may be blocked waiting for the task to finish. We can retry the task, but assuming it keeps killing workers, we should at some point give up and raise a python exception in the blocked processes.

**Comment by robertnishihara** [comment]
Copying the example from #397, this is a concrete example for testing this functionality.

```python
import ray

ray.init(redirect_output=True)

@ray.remote
def f():
  eval("exit()")

ray.get(f.remote())
```

This script will currently hang.

**Comment by robertnishihara** [comment]
Addressed by #419.
